src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc

```diff
@@ -38,625 +39,937 @@
-        j                    = std::max( std::min( j, N - 1 ), zero );
-        i                    = std::max( std::min( i, N - 1 ), zero );
-        return idx2st( t * N * N + j * N + i );
-    };
-
-    // ---------------------------------------------------------------------------
-    // 2. GLOBAL CELL DISTRIBUTION
-    // ---------------------------------------------------------------------------
-
-    // Define cell record.
-    // This currently keeps track of more information than we probably need.
-    // This may be reduced once we've implemented halos.
-    struct CellRecord {
-        gidx_t globalIdx{undefinedGlobalIdx};  // Global ID.
-        idx_t localIdx{undefinedIdx};          // Local ID.
-        idx_t part{undefinedIdx};              // Partition.
-        PointXY xy{};                          // Position.
-    };
-
-    // Define ij bounding box for each face (this partition only).
-    struct BoundingBox {
-        idx_t iBegin{std::numeric_limits<idx_t>::max()};
-        idx_t iEnd{std::numeric_limits<idx_t>::min()};
-        idx_t jBegin{std::numeric_limits<idx_t>::max()};
-        idx_t jEnd{std::numeric_limits<idx_t>::min()};
-    };
-
-    // Make list of all cells.
-    auto globalCells = std::vector<CellRecord>( idx2st( nCells ) );
-
-    // Initialise bounding box.
-    auto cellBounds = std::vector<BoundingBox>( idx2st( nTiles ) );
-
-    // Set xy and tji grid iterators.
-    auto tjiIt = csGrid.tij().begin();
-    auto xyIt  = csGrid.xy().begin();
-
-    // Set counters for cell local indices.
-    auto cellIdxCount = std::vector<idx_t>( nParts, 0 );
-
-    for ( gidx_t gridIdx = 0; gridIdx < csGrid.size(); ++gridIdx ) {
-        // Grid (t, j, i) order does not have to match mesh (t, j, i), although
-        // in practice it probably does.
-
-        // Get cell index.
-        const auto t       = ( *tjiIt ).t();
-        const auto j       = ( *tjiIt ).j();
-        const auto i       = ( *tjiIt ).i();
-        const auto cellIdx = getCellIdx( t, j, i );
-        auto& cell         = globalCells[cellIdx];
-
-        // Set global index.
-        cell.globalIdx = gridIdx + 1;
-
-        // Set partition and remote index.
-        cell.part = distribution.partition( gridIdx );
-
-        // Set local index.
-        cell.localIdx = cellIdxCount[idx2st( cell.part )]++;
-
-        // Set xy
-        cell.xy = *xyIt;
-
-        if ( cell.part == st2idx( thisPart ) ) {
-            // Keep track of local (t, j, i) bounds.
-            auto& bounds  = cellBounds[idx2st( t )];
-            bounds.iBegin = std::min( bounds.iBegin, i );
-            bounds.iEnd   = std::max( bounds.iEnd, i + 1 );
-            bounds.jBegin = std::min( bounds.jBegin, j );
-            bounds.jEnd   = std::max( bounds.jEnd, j + 1 );
-        }
-        // Increment iterators.
-        ++tjiIt;
-        ++xyIt;
+void CubedSphereMeshGenerator::generate_mesh(const CubedSphereGrid& csGrid,
```

Comment:
Wow. This method is extremely well structured and commented. I don't pretend to understand it all but I can tell it contains a tremendous amount of work, detail, and complexity.

Even as well-organized and documented as it is, though, it still makes me nervous that it is over 800 lines long. That is significantly longer than most most recommended guidelines I've heard as a software engineer, in order to keep functions easy for others (or an older version of the original programmer) to understand and maintain. But I also know that those are just guidelines and perhaps don't apply to this situation. So I guess that's just something I'd think about, to see if you think some sort of refactoring into smaller functions makes sense. Perhaps not for this PR, but maybe while all this detail is still fresh in your head.

Reply:
I completely agree. It's way longer than I'm happy with. Many of the mesh generators in that directory sprawl a bit, and it was an absolute pain to debug. If I had the time, I'd refactor into a set of distinct classes. Unfortunately I don't.